File: src/backend/home_api/games/game/models/Paddle.py
from games.game.models.Ball import Ball
from games.game.constants import FIELD_HEIGHT

class Paddle:
	def __init__(self, x, y, width, height, speed):
		"""
		:param x: int (x position)
		:param y: int (y position)
		:param width: int
		:param height: int
		:param vy: int (y velocity)
		"""
		self.default_x = x
		self.default_y = y
		self.default_width = width
		self.default_height = height
		self.default_speed = speed
		self.default_vy = 0

		self.x = self.default_x
		self.y = self.default_y
		self.width = self.default_width
		self.height = self.default_height
		self.speed = self.default_speed
		self.vy = self.default_vy

	def move(self, dt):
		"""
		Update the paddle's position based on its velocity (`vy`).
		Ensure it does not go out of bounds.
		"""
		self.y += self.vy * dt
		# Check top boundary
		if self.y - self.height / 2 < -FIELD_HEIGHT / 2:
			self.y = -FIELD_HEIGHT / 2 + self.height / 2
		# Check bottom boundary
		elif self.y + self.height / 2 > FIELD_HEIGHT / 2:
			self.y = FIELD_HEIGHT / 2 - self.height / 2

	def move_up(self, is_pressed):
		if is_pressed:
			self.vy = -self.speed
		else:
			self.vy = 0 if self.vy < 0 else self.vy # prevent erase down movement
		
	def move_down(self, is_pressed):
		if is_pressed:
			self.vy = self.speed
		else:
			self.vy = 0 if self.vy > 0 else self.vy # prevent erase up movement

	def to_dict(self):
		return {
			'x': self.x,
			'y': self.y,
			'width': self.width,
			'height': self.height,
			'vy': self.vy
		}
	
	def reset(self):
		self.x = self.default_x
		self.y = self.default_y
		self.width = self.default_width
		self.height = self.default_height
		self.vy = self.default_vy

File: src/frontend/src/js/pages/play/local/_components/Paddle.js
import { FIELD_HEIGHT } from "../LocalRunner.js";

export class Paddle {
	constructor(x, y, width, height, speed) {
		this.default_x = x;
		this.default_y = y;
		this.default_width = width;
		this.default_height = height;
		this.default_speed = speed;
		this.default_vy = 0;

		this.x = this.default_x;
		this.y = this.default_y;
		this.width = this.default_width;
		this.height = this.default_height;
		this.speed = this.default_speed;
		this.vy = this.default_vy;
	}

	move(dt) {
		this.y += this.vy * dt;

		if (this.y - this.height / 2 < -FIELD_HEIGHT / 2) {
			this.y = -FIELD_HEIGHT / 2 + this.height / 2;
		} else if (this.y + this.height / 2 > FIELD_HEIGHT / 2) {
			this.y = FIELD_HEIGHT / 2 - this.height / 2;
		}
	}

	moveUp(is_pressed) {
		if (is_pressed) {
			this.vy = -this.speed;
		} else {
			this.vy = this.vy < 0 ? 0 : this.vy;
		}
	}

	moveDown(is_pressed) {
		if (is_pressed) {
			this.vy = this.speed;
		} else {
			this.vy = this.vy > 0 ? 0 : this.vy;
		}
	}

	reset() {
		this.x = this.default_x;
		this.y = this.default_y;
		this.width = this.default_width;
		this.height = this.default_height;
		this.speed = this.default_speed;
		this.vy = this.default_vy;
	}

	to_json() {
		return {
			x: this.x,
			y: this.y,
			width: this.width,
			height: this.height,
			vy: this.vy,
		};
	}
}